File: macos/AirNinjaCore/Sources/AirNinjaCore/DeviceId.swift
import Foundation

public enum DeviceId {
    private static let length = 52

    public static func fromPublicKey(_ staticPublicKey: Data) -> String {
        precondition(!staticPublicKey.isEmpty, "Static public key must not be empty")
        let digest = Sha256.hash(staticPublicKey)
        return String(Base32.encodeLowerNoPadding(digest).prefix(length))
    }
}

File: protocol-core/src/main/kotlin/com/airninja/protocol/identity/DeviceId.kt
package com.airninja.protocol.identity

import com.airninja.protocol.crypto.Base32
import com.airninja.protocol.crypto.Sha256

object DeviceId {
    private const val LENGTH = 52

    fun fromPublicKey(staticPublicKey: ByteArray): String {
        require(staticPublicKey.isNotEmpty()) { "Static public key must not be empty" }
        return Base32.encodeLowerNoPadding(Sha256.hash(staticPublicKey)).take(LENGTH)
    }
}
